### PR TITLE
OSAC-1110: StorageTier API

### PR DESCRIPTION
**Pull request by rgolangh**
## [OSAC-1110](https://redhat.atlassian.net/browse/OSAC-1110): StorageTier API

**Jira:** https://redhat.atlassian.net/browse/OSAC-1110
**Design:** https://github.com/osac-project/enhancement-proposals/pull/75

### Summary

Implements the StorageTier private gRPC API. StorageTier is the catalog layer that lets Cloud Provider Admins define named storage offerings (e.g., "fast", "standard", "archive") backed by registered StorageBackends with QoS properties (bandwidth, quota, encryption). Downstream, Tenant Storage Onboarding ([OSAC-23](https://redhat.atlassian.net/browse/OSAC-23)) consumes these tiers to determine what storage to provision.

### Why

OSAC needs a way to represent differentiated storage levels so operators can offer tiered storage to tenants. Without this, all tenants get the same storage characteristics and there is no mechanism for downstream workflows to determine which StorageClasses to create or what QoS policies to apply.

### Design decisions

- **DB-backed, no CRD**: follows the NetworkClass pattern. Tiers are platform-scoped (SharedTenant), stored in PostgreSQL, and have no reconciler.
- **v0.1 single-backend constraint**: exactly one backend association per tier. The schema supports multiple backends for future milestones, but the server enforces one for now.
- **Bidirectional referential integrity via triggers**: a materialized helper table (`storage_tier_backends`) mirrors backend references from the JSONB data column, enabling FOR SHARE locking on create/update and deletion protection on the backend side. This covers race conditions that Go-level validation alone cannot.
- **Backend existence validated at two layers**: the server checks via DAO (user-friendly gRPC errors), and the DB trigger catches races (safety net).

### Testing

**Server tests** cover full CRUD lifecycle, list with pagination/filtering/ordering, partial updates via field mask, backend replacement, optimistic locking with stale version, name uniqueness (including reuse after delete), name and tenant immutability, and validation edge cases (nil object, missing name, missing/empty/non-existent backends, multiple backends, clearing backends on update).

**Migration tests** cover table creation, tenant FK enforcement, name uniqueness, column immutability, helper table materialization on insert, rejection of non-existent and soft-deleted backend references, deletion protection with tier count reporting, and cleanup scenarios (unused backend deletion, delete-tier-then-delete-backend).

### CODE REVIEW

**Comment by akshaynadkarni** [comment]
@rgolangh Just sharing the overall summary. I will do a more thorough review next week. 


| # | Type | Severity | Confidence | File | Description | Link |
|---|------|----------|------------|------|-------------|------|
| 1 | Issue | Medium | 95% | `start_grpc_server_cmd.go` | StorageBackendsDAO not wired in; server-level backend validation is dead code | [L959-L965](https://github.com/osac-project/fulfillment-service/blob/6d21594520c6dbf43dd432559c847d0bf2af6731/internal/cmd/service/start/grpcserver/start_grpc_server_cmd.go#L959-L965) |
| 2 | Suggestion | Low | 80% | `private_storage_tiers_server.go` | Duplicate backend validation logic in create and update paths | [L204-L220](https://github.com/osac-project/fulfillment-service/blob/6d21594520c6dbf43dd432559c847d0bf2af6731/internal/servers/private_storage_tiers_server.go#L204-L220) |
| 3 | Question | Medium | 65% | `private_storage_tiers_server.go` | Update path allows clearing all backends (empty list passes validation) | [L233-L234](https://github.com/osac-project/fulfillment-service/blob/6d21594520c6dbf43dd432559c847d0bf2af6731/internal/servers/private_storage_tiers_server.go#L233-L234) |
| 4 | Suggestion | Low | 85% | `private_storage_tiers_server_test.go` | No unit tests exercise server-level backend existence validation | [test file](https://github.com/osac-project/fulfillment-service/blob/6d21594520c6dbf43dd432559c847d0bf2af6731/internal/servers/private_storage_tiers_server_test.go) |

**Comment by akshaynadkarni** [comment]
Nit: description says migration 72/73 but the files are 74/75. Also references a `setPayload()` change in `generic_server.go` that isn't in the diff. Looks like stale text from before rebasing.

File: internal/servers/private_storage_tiers_server.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+func (s *PrivateStorageTiersServer) validateStorageTierUpdate(ctx context.Context,
+	newST *privatev1.StorageTier, existingST *privatev1.StorageTier) error {
+
+	if newST.GetMetadata() != nil && newST.GetMetadata().GetName() != "" &&
+		newST.GetMetadata().GetName() != existingST.GetMetadata().GetName() {
+		return grpcstatus.Errorf(grpccodes.InvalidArgument,
+			"field 'metadata.name' is immutable and cannot be changed from '%s' to '%s'",
+			existingST.GetMetadata().GetName(), newST.GetMetadata().GetName())
+	}
+	backends := newST.GetBackends()
```
Comment (akshaynadkarni) [inline internal/servers/private_storage_tiers_server.go:218]: Create requires at least one backend, but this check skips validation entirely when `backends` is empty. If the field mask includes `"backends"` with an empty list, the tier ends up with zero associations. Is that intentional?

File: internal/servers/private_storage_tiers_server.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	}
+	backends := st.GetBackends()
+	if len(backends) == 0 {
+		return grpcstatus.Errorf(grpccodes.InvalidArgument, "field 'backends' is required and must not be empty")
+	}
+	if len(backends) > 1 {
+		return grpcstatus.Errorf(grpccodes.InvalidArgument,
+			"only one backend association is supported in v0.1, but %d were provided", len(backends))
+	}
+	for _, ba := range backends {
```
Comment (akshaynadkarni) [inline internal/servers/private_storage_tiers_server.go]: Nit: this validation loop is duplicated in `validateStorageTierUpdate` (L239-255). Consider extracting a shared helper.

**Comment by akshaynadkarni** [comment]
@coderabbitai review

**Comment by akshaynadkarni** [comment]
@coderabbitai review

File: internal/cmd/cli/create/storagebackend/create_storagebackend_cmd.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		"username",
+		"",
+		usernameFlagHelp,
+	)
+	flags.StringVar(
+		&runner.password,
+		"password",
+		"",
+		passwordFlagHelp,
+	)
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/cmd/cli/create/storagebackend/create_storagebackend_cmd.go:69]: The `login` command uses the same `--password` flag pattern. This is a codebase-wide convention, not something to address in this PR.

File: internal/database/migrations/75_create_storage_tiers_tables_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+			"changed-name", "immutable-id",
+		)
+		Expect(err).To(HaveOccurred())
+
+		_, err = conn.Exec(ctx,
+			`update storage_tiers set tenant = $1 where id = $2`,
+			"other-tenant", "immutable-id",
+		)
+		Expect(err).To(HaveOccurred())
+	})
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/database/migrations/75_create_storage_tiers_tables_test.go:96]: The `storage_backends` migration test (migration 62) follows the same pattern: only tests `name` and `tenant`, not `id`. Changing `id` is already prevented by the primary key constraint. Consistent with existing tests.

File: internal/database/migrations/75_create_storage_tiers_tables.up.sql [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  data jsonb not null,
+  version integer not null default 0
+);
+
+create index storage_tiers_by_creator on storage_tiers (creator);
+create index storage_tiers_by_label on storage_tiers using gin (labels);
+
+-- Name uniqueness across all tiers (active and soft-deleted).
+create unique index storage_tiers_unique_name
+  on storage_tiers (name);
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/database/migrations/75_create_storage_tiers_tables.up.sql:52]: `storage_backends` uses the identical non-partial unique index. StorageTier has no finalizers, so deletes archive immediately and the name is freed. The test "Create after delete of same name succeeds" confirms this works.

File: internal/database/migrations/76_add_storage_tier_ref_triggers_test.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = conn.Exec(ctx,
+			`update storage_tiers set deletion_timestamp = now() where id = 'tier-gone'`)
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = conn.Exec(ctx,
+			`update storage_backends set deletion_timestamp = now() where id = 'sb-was-ref'`)
+		Expect(err).ToNot(HaveOccurred())
+	})
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/database/migrations/76_add_storage_tier_ref_triggers_test.go:273]: No other materialization trigger test in the codebase covers the UPDATE path. This would be a new testing standard, not a gap specific to this PR.

File: internal/database/migrations/76_add_storage_tier_ref_triggers.up.sql [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  bid text;
+begin
+  delete from storage_tier_backends where storage_tier_id = new.id;
+
+  for bid in
+    select jsonb_array_elements(new.data->'backends')->>'backend_id'
+  loop
+    insert into storage_tier_backends (storage_tier_id, backend_id)
+    values (new.id, bid);
+  end loop;
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/database/migrations/76_add_storage_tier_ref_triggers.up.sql:49]: The v0.1 server validation enforces exactly one backend per tier, so duplicate backend_ids cannot reach the trigger through the API. No other materialization trigger in the codebase uses `DISTINCT`.

File: internal/database/migrations/75_create_storage_tiers_tables_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/database/migrations/75_create_storage_tiers_tables_test.go:96]: StorageTier is platform-scoped (SharedTenant). The `project` column is part of the generic DAO schema template but is never set by the server. It will always be the empty default. Testing immutability of a column that is never written to does not add meaningful coverage.

File: internal/servers/private_storage_tiers_server_test.go [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+				}.Build(),
+				UpdateMask: &fieldmaskpb.FieldMask{Paths: []string{"backends"}},
+			}.Build())
+			Expect(err).ToNot(HaveOccurred())
+			Expect(updateResponse.GetObject().GetBackends()).To(HaveLen(1))
+			Expect(updateResponse.GetObject().GetBackends()[0].GetProtocol()).To(Equal(
+				privatev1.StorageProtocol_STORAGE_PROTOCOL_BLOCK))
+			Expect(updateResponse.GetObject().GetBackends()[0].GetMaxReadBandwidthMbs()).To(Equal(int32(2000)))
+			Expect(updateResponse.GetObject().GetBackends()[0].GetQuotaGib()).To(Equal(int64(2048)))
+		})
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/servers/private_storage_tiers_server_test.go:323]: Addressed in 0776d4e. Added "Update clearing backends fails" test.

File: internal/servers/private_storage_tiers_server_test.go [hunk 9]
```diff
@@ -0,0 +1,664 @@
...
+							Tenant: "other-tenant",
+						}.Build(),
+					}.Build(),
+					UpdateMask: &fieldmaskpb.FieldMask{Paths: []string{"metadata.tenant"}},
+				}.Build())
+				Expect(err).To(HaveOccurred())
+				st, ok := status.FromError(err)
+				Expect(ok).To(BeTrue())
+				Expect(st.Code()).ToNot(Equal(codes.OK))
+			})
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/servers/private_storage_tiers_server_test.go:614]: Addressed in 0776d4e. Changed to `Equal(codes.InvalidArgument)` + `ContainSubstring("immutable")`.

File: internal/servers/private_storage_tiers_server.go [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+		return
+	}
+
+	result = &PrivateStorageTiersServer{
+		logger:             b.logger,
+		generic:            generic,
+		storageBackendsDAO: b.storageBackendsDAO,
+	}
+	return
+}
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/servers/private_storage_tiers_server.go:117]: Addressed in 0776d4e. DAO is now mandatory in `Build()` and nil guards are removed.

File: internal/servers/private_storage_tiers_server.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+			if err != nil {
+				var notFoundErr *dao.ErrNotFound
+				if errors.As(err, &notFoundErr) {
+					return grpcstatus.Errorf(grpccodes.NotFound,
+						"storage backend with identifier '%s' not found", ba.GetBackendId())
+				}
+				return grpcstatus.Errorf(grpccodes.Internal,
+					"failed to validate storage backend '%s'", ba.GetBackendId())
+			}
+		}
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/servers/private_storage_tiers_server.go]: Addressed in 0776d4e. Extracted shared `validateBackends` helper.

File: internal/servers/private_storage_tiers_server.go [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+					if errors.As(err, &notFoundErr) {
+						return grpcstatus.Errorf(grpccodes.NotFound,
+							"storage backend with identifier '%s' not found", ba.GetBackendId())
+					}
+					return grpcstatus.Errorf(grpccodes.Internal,
+						"failed to validate storage backend '%s'", ba.GetBackendId())
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (akshaynadkarni) [reply internal/servers/private_storage_tiers_server.go:255]: Addressed in 0776d4e. Update now rejects empty backends when `"backends"` is in the field mask.

**Comment by akshaynadkarni** [comment]
@coderabbitai resolve

**Review by akshaynadkarni — APPROVED** [review]
LGTM
